File: backend_py_old/routes/penalties.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from ..database import get_db
from ..models import Player, DailyLog, Quest
from datetime import date, timedelta

router = APIRouter()

@router.post("/penalties/apply")
def apply_penalties(db: Session = Depends(get_db)):
    # Run for yesterday's logs if not already applied
    yesterday = date.today() - timedelta(days=1)
    logs = db.query(DailyLog).filter(DailyLog.date == yesterday).all()
    
    if not logs:
        # If no logs exist for yesterday, check if we should have some
        # This usually happens if the user didn't open the app yesterday
        # We'll create them now to apply penalties
        quests = db.query(Quest).filter(Quest.is_active == True).all()
        for q in quests:
            new_log = DailyLog(date=yesterday, quest_id=q.id, completed=False)
            db.add(new_log)
        db.commit()
        logs = db.query(DailyLog).filter(DailyLog.date == yesterday, DailyLog.penalty_applied == False).all()
    else:
        logs = [l for l in logs if not l.penalty_applied]

    player = db.query(Player).first()
    penalty_summary = {"missed": 0, "xp_lost": 0, "stats_lost": {}}
    
    all_completed = True
    for log in logs:
        if not log.completed:
            all_completed = False
            quest = db.query(Quest).filter(Quest.id == log.quest_id).first()
            
            # XP Penalty
            xp_loss = abs(quest.penalty_xp)
            player.total_xp = max(0, player.total_xp - xp_loss)
            penalty_summary["xp_lost"] += xp_loss
            
            # Stat Penalty
            stat_loss = abs(quest.penalty_stat)
            if stat_loss > 0:
                cur_stat = getattr(player, quest.stat_key)
                setattr(player, quest.stat_key, max(1, cur_stat - stat_loss))
                penalty_summary["stats_lost"][quest.stat_key] = penalty_summary["stats_lost"].get(quest.stat_key, 0) + stat_loss
            
            log.penalty_applied = True
            penalty_summary["missed"] += 1
        else:
            log.penalty_applied = True # Marked as processed

    if all_completed and len(logs) > 0:
        player.streak_days += 1
        player.total_xp += 25 # Streak bonus
        penalty_summary["streak_bonus"] = 25
    elif not all_completed:
        player.streak_days = 0

    db.commit()
    return penalty_summary

File: backend/routes/penalties.js
import express from 'express';
import { Player, DailyLog, Quest } from '../models/models.js';

const router = express.Router();

router.post('/penalties/apply', async (req, res) => {
  try {
    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    const yesterdayStr = yesterday.toISOString().split('T')[0];

    let logs = await DailyLog.findAll({ where: { date: yesterdayStr } });
    
    if (logs.length === 0) {
      const activeQuests = await Quest.findAll({ where: { is_active: true } });
      const newLogs = activeQuests.map(q => ({
        date: yesterdayStr,
        quest_id: q.id,
        completed: false
      }));
      await DailyLog.bulkCreate(newLogs);
      logs = await DailyLog.findAll({ where: { date: yesterdayStr } });
    }

    const unprocessedLogs = logs.filter(l => !l.penalty_applied);
    const player = await Player.findOne();
    const penaltySummary = { missed: 0, xp_lost: 0, stats_lost: {} };
    
    let allCompleted = true;
    for (const log of unprocessedLogs) {
      if (!log.completed) {
        allCompleted = false;
        const quest = await Quest.findByPk(log.quest_id);
        
        const xpLoss = Math.abs(quest.penalty_xp);
        player.total_xp = Math.max(0, player.total_xp - xpLoss);
        penaltySummary.xp_lost += xpLoss;

        const statLoss = Math.abs(quest.penalty_stat);
        if (statLoss > 0) {
          const statKey = quest.stat_key;
          player[statKey] = Math.max(1, (player[statKey] || 0) - statLoss);
          penaltySummary.stats_lost[statKey] = (penaltySummary.stats_lost[statKey] || 0) + statLoss;
        }
        
        log.penalty_applied = true;
        penaltySummary.missed += 1;
        await log.save();
      } else {
        log.penalty_applied = true;
        await log.save();
      }
    }

    if (allCompleted && logs.length > 0) {
      player.streak_days += 1;
      player.total_xp += 25; // Streak bonus
      penaltySummary.streak_bonus = 25;
    } else if (!allCompleted) {
      player.streak_days = 0;
    }

    await player.save();
    res.json(penaltySummary);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;
